src/function/gds/output_writer.cpp

```diff
@@ -82,30 +84,32 @@ void PathsOutputWriter::write(processor::FactorizedTable& fTable, nodeID_t dstNo
         auto top = curPath[curPath.size() - 1];
         auto topNodeID = top->getNodeID();
         if (topNodeID == sourceNodeID) {
-            beginWritingNewPath(curPath.size());
-            if (extendFromSource) {
-                // Write path in reverse direction because we append ParentList from dst to src.
-                for (auto i = 1u; i < curPath.size(); i++) {
-                    auto p = curPath[curPath.size() - 1 - i];
-                    addNode(p->getNodeID(), i - 1);
-                    addEdge(p->getEdgeID(), p->isFwdEdge(), i);
+            // check path
+            if (checkPathNode(curPath)) {
```

Comment:
I think this is not the efficient way to apply the node predicate. Your algorithm is: enumerate every path and then check if all the nodes in it pass the predicate. Instead, you should do it while running the DFS logic in the path construction, so in this PR I would apply it inside 140-147 branch:
```
...
else {
            auto parent = bfsGraph.getInitialParentAndNextPtr(topNodeID);
            while (parent->getIter() != top->getIter() - 1 && checkNodePredicateIfIterNotOne(parent)) {
                parent = parent->getNextPtr();
            }
            if (parent != null) { 
               curPath.push_back(parent);
               backtracking = false;
           }
        }
```
First I am assuming that there is a`checkNodePredicateIfIterNotOne` function that checks that the parent is not at iteration 1, in which case we don't need to check if a node predicate applies to it, because it is guaranteed to be the source, and then applies the node predicate (if any) if the iteration is not 1. Second, now you need the `if (parent != null)` condition because now there is a chance that the parent is null. before you could guarantee that some parent must always satisfy this condition: parent->getIter() != top->getIter() - 1.

Further lines 129-137 also need to change (the if condition there needs to be a while loop now):
```
            while (top->getNextPtr() != nullptr  &&
                ((curPath.size() == 1 && top->getNextPtr()->getIter() >= lowerBound) ||
                    top->getNextPtr()->getIter() == top->getIter())) {
                if (checkNodePredicateIfIterNotOne(top->getNextPtr()) {
                  curPath[curPath.size() - 1] = top->getNextPtr();
                  backtracking = false;
                  if (curPath.size() == 1) {
                     lengthVector->setValue<int64_t>(0, curPath[0]->getIter());
                  }
                  break;
                }
            }
            if (backtracking) {
                curPath.pop_back();
            }
```
Something like this. The if condition needs to be replaced with a while loop because we need to check the next parent, which may be at the right level but not pass the node predicate. If it does not pass the node predicate (the inner `if (checkNodePredicateIfIterNotOne(top->getNextPtr())` branch, we should continue to see if there are other parents who are the right level. The previous else branch (currently on line 137) now needs to be an if condition that determines whether or not the while loop was able to find an appropriate parent.

The backtracking logic is hard to understand. So you should have test cases that tests all of these edge cases. This is very important because it will be a nightmare to debug these unless we make sure that we test every single line of this path outputting logic. For example, there should be a test case that tests that each of the while loops in this complex logic iterates multiple times.  Let's talk about your test cases.

Also: if you don't want to do this optimization now, then you should open an issue about it. Though I strongly, strongly suggest we do it now, it's the right time to do all these optimizations and be done with them. The next person will have harder time doing them I think.

Reply:
As we discussed, I'll do this in a separate PR. And I have updated the issue to track this.